modules/twinkleblock.js

```diff
@@ -430,27 +430,67 @@ Twinkle.block.callback.change_action = function twinkleblockCallbackChangeAction
 
 		$form.find('[name=pagerestrictions]').select2({
 			width: '100%',
-			tags: true,
-			placeholder: 'Enter pages to block user from',
+			placeholder: 'Select pages to block user from',
 			maximumSelectionLength: 10, // Software limitation [[phab:T202776]]
-			language: {
-				noResults: function() {
-					return 'No pages entered yet';
+			minimumInputLength: 1, // prevent ajax call when empty
+			ajax: {
+				url: mw.util.wikiScript('api'),
+				dataType: 'json',
+				delay: 100,
+				data: function(params) {
+					var title = mw.Title.newFromText(params.term);
+					if (!title) {
+						return;
+					}
+					return {
+						'action': 'query',
+						'format': 'json',
+						'list': 'allpages',
+						'apfrom': title.title,
+						'apnamespace': title.namespace,
+						'aplimit': '10'
+					};
+				},
+				processResults: function(data) {
+					return {
+						results: data.query.allpages.map(function(page) {
+							var title = mw.Title.newFromText(page.title, page.ns).toText();
+							return {
+								id: title,
+								text: title
+							};
+						})
+					};
 				}
+			},
+			templateSelection: function(choice) {
+				return $('<a>').text(choice.text).attr({
+					href: mw.util.getUrl(choice.text),
+					target: '_blank'
+				});
 			}
 		});
 
 
 		$form.find('[name=namespacerestrictions]').select2({
 			width: '100%',
+			matcher: function(params, data) {
+				var originalMatcher = $.fn.select2.defaults.defaults.matcher;
+				var result = originalMatcher(params, data);
+				if (!params.term || (result &&
+					new RegExp('\\b' + mw.util.escapeRegExp(params.term), 'i').test(result.text))) {
+					return result;
+				}
+				return null;
+			},
 			language: {
 				searching: Morebits.select2.queryInterceptor
 			},
 			templateResult: Morebits.select2.highlightSearchMatches,
-			placeholder: 'Select namespaces to block user from'
+			placeholder: 'Select namespaces to block user from',
+			closeOnSelect: false
```

Comment:
Not entirely sure about this.  It's definitely a positive and makes things easier if you're using the cursor, but for someone using the keyboard, I think it's a bit of a step backward.  The text you entered to find a match lingers after you've selected that option: that means that you have to delete it before typing your next namespace.  It also means the best way to close the menu is to hit tab, or enter if you're on an option you already selected, which isn't particularly intuitive.

My instinct is that we're likely not to see many multiple-namespace blocks, or at least it won't be the norm, so could probably be left off.